src/ledger/LedgerManagerImpl.cpp

```diff
@@ -2791,4 +2835,299 @@ LedgerManagerImpl::ApplyState::addAnyContractsToModuleCache(
+        {
+            applySorobanStage(mApp.getAppConnector(), header,
+                              *preExecutionGlobalState, stage,
+                              sorobanBasePrngSeed);
+        }
+
+        // Build parallel results for comparison with just the Soroban subset
+        TransactionResultSet sequentialSorobanResults;
+        TransactionResultSet parallelTxResultSet;
+
+        sequentialSorobanResults.results.reserve(newSorobanTxs.size());
+        parallelTxResultSet.results.reserve(newSorobanTxs.size());
+
+        for (size_t i = 0; i < newSorobanTxs.size(); ++i)
+        {
+            auto const& tx = newSorobanTxs[i];
+
+            // Get the sequential result for this transaction
+            size_t originalIndex = sorobanStartIndex + i;
+            auto const& seqResult =
+                originalTxResultSet.results.at(originalIndex);
+            sequentialSorobanResults.results.emplace_back(seqResult);
+
+            // Get the parallel result
+            auto const& mutableResult = parallelMutableResults[i];
+
+            TransactionResultPair resultPair;
+            resultPair.transactionHash = tx->getContentsHash();
+            resultPair.result = mutableResult->getXDR();
+
+            parallelTxResultSet.results.emplace_back(resultPair);
+        }
+
+        // Compare results between sequential and parallel execution
+        compareParallelExecutionResults(sequentialSorobanResults,
+                                        parallelTxResultSet);
+
+        CLOG_INFO(Ledger, "Parallel re-execution test completed successfully");
+    }
+    catch (std::exception const& e)
+    {
+        CLOG_ERROR(Ledger, "Exception during parallel re-execution test: {}",
+                   e.what());
+    }
+}
+void
+LedgerManagerImpl::compareParallelExecutionResults(
+    TransactionResultSet const& sequential,
+    TransactionResultSet const& parallel)
+{
+    if (sequential.results.size() != parallel.results.size())
+    {
+        CLOG_WARNING(Ledger,
+                     "Result count mismatch: sequential={}, parallel={}",
+                     sequential.results.size(), parallel.results.size());
+        return;
+    }
+
+    size_t differences = 0;
+    for (size_t i = 0; i < sequential.results.size(); ++i)
+    {
+        auto const& seqResult = sequential.results[i];
+        auto const& parResult = parallel.results[i];
+
+        if (seqResult.transactionHash != parResult.transactionHash)
+        {
+            CLOG_ERROR(Ledger, "Transaction hash mismatch at index {}", i);
+            differences++;
+            continue;
+        }
+
+        bool resultsDiffer = false;
+
+        // Compare result codes
+        if (seqResult.result.result.code() != parResult.result.result.code())
+        {
+            resultsDiffer = true;
+        }
+        // For successful transactions, compare inner results
+        else if (seqResult.result.result.code() == txSUCCESS)
```

Comment:
This whole branch doesn't make much sense to me (like why is txSUCCESS special). It should be ok to just compare the XDR and log diffs via xdr_to_string.

Reply:
thrown out and rewrote